clang/lib/Sema/SemaDecl.cpp

```diff
@@ -2380,7 +2380,7 @@ FunctionDecl *Sema::CreateBuiltin(IdentifierInfo *II, QualType Type,
   }
 
   FunctionDecl *New = FunctionDecl::Create(Context, Parent, Loc, Loc, II, Type,
-                                           /*TInfo=*/nullptr, SC_Extern,
+                                           /*TInfo=*/nullptr, SC_None,
```

Comment:
This change looks suspicious to me -- this is creating a builtin function, which should be modeled as an extern. In C++, you seem to be relying on the language linkage being sufficient for that, but C has no notion of language linkage and so I worry that the declaration for the builtin will be incorrectly handled in C.

Reply:
You are right, I also paid attention to that issue. However, when a storage class specifier is omitted for a function, it's storage class is implicitly supposed to be `extern`. Thus, it's fine.

Furthermore, the meaning of this field according to the documentation comments is not for the actual storage duration or linkage, but for the storage-class specifier _as present in the source code_. Since builtins aren't really present in the source code, it seems further reasonable to omit a specifier.

Reply:
> You are right, I also paid attention to that issue. However, when a storage class specifier is omitted for a function, it's storage class is implicitly supposed to be extern. Thus, it's fine.

That's not actually correct -- the declaration of a function at block scope should definitely *not* be extern, it should have no linkage: https://godbolt.org/z/81fMaPaTq

> Furthermore, the meaning of this field according to the documentation comments is not for the actual storage duration or linkage, but for the storage-class specifier as present in the source code. Since builtins aren't really present in the source code, it seems further reasonable to omit a specifier.

Builtins should behave as-if they were declared in a header file that was included in the TU, and so they typically would be marked as `extern`.

Reply:
> That's not actually correct -- the declaration of a function at block scope should definitely _not_ be extern, it should have no linkage: https://godbolt.org/z/81fMaPaTq

I may be wrong to rely on cppreference, but the following page tells that indeed a function at a block scope has implicit `extern`, which could also be made explicit: https://en.cppreference.com/w/c/language/storage_duration.

I also checked your example locally by compiling (via clang) `main.c` with an additional `defs.c`: 
main.c:
```c
int main(void) {
  void f();
  extern void g();

  f();
  g();

  return 0;
}
```

defs.c: 
```c
#include <stdio.h>

void f() {
  printf("f\n");
}

void g() {
  printf("g\n");
}

```

The executable works as expected. 

I will also take a look at the C99 standard draft.

Reply:
> > That's not actually correct -- the declaration of a function at block scope should definitely _not_ be extern, it should have no linkage: https://godbolt.org/z/81fMaPaTq
> 
> I may be wrong to rely on cppreference, but the following page tells that indeed a function at a block scope has implicit `extern`, which could also be made explicit: https://en.cppreference.com/w/c/language/storage_duration.
> 
> I also checked your example locally by compiling (via clang) `main.c` with an additional `defs.c`: main.c:
> 
> ```c
> int main(void) {
>   void f();
>   extern void g();
> 
>   f();
>   g();
> 
>   return 0;
> }
> ```
> 
> defs.c:
> 
> ```c
> #include <stdio.h>
> 
> void f() {
>   printf("f\n");
> }
> 
> void g() {
>   printf("g\n");
> }
> ```
> 
> The executable works as expected.
> 
> I will also take a look at the C99 standard draft.

Paragraph 6.2.2.5 says: 

> If the declaration of an identifier for a function has no storage-class specifier, its linkage
is determined exactly as if it were declared with the storage-class specifier extern. If
the declaration of an identifier for an object has file scope and no storage-class specifier,
its linkage is external.

Thus, seemingly confirming that a function declared at the block scope is declared with an implicit `extern`.

Reply:
Ah, you're right, it's only *objects* declared at local scope that have no linkage (the following paragraph), not *functions*.

Reply:
I think I've come around to being okay with these changes, thank you!